Fixes #9391 - port/move Jetty WebSocket APIs, client and server to je…

sbordet opened the pull request:
…tty-core.

* Moved -api, -common, -client to jetty-core/jetty-websocket.
* Implemented jetty-core/jetty-websocket/jetty-websocket-jetty-server using only Jetty core APIs, not Servlet.
* Fixed Graceful shutdown order.
* Fixed mistakes in HttpFieldsWrapper, wrongly calling HttpHeader.name() instead of asString().
* Updated tests to pass cleanly.
* Fixed BOMs and POM dependencies.
* Introduced websocket-jetty.mod and websocket-jetty-client.mod, now used by ee10's Jetty WebSocket.
* Fixed OSGi references to old artifactIds.

sbordet committed:
Fixes #9391 - port/move Jetty WebSocket APIs, client and server to jetty-core.

* Moved -api, -common, -client to jetty-core/jetty-websocket.
* Implemented jetty-core/jetty-websocket/jetty-websocket-jetty-server using only Jetty core APIs, not Servlet.
* Fixed Graceful shutdown order.
* Fixed mistakes in HttpFieldsWrapper, wrongly calling HttpHeader.name() instead of asString().
* Updated tests to pass cleanly.
* Fixed BOMs and POM dependencies.
* Introduced websocket-jetty.mod and websocket-jetty-client.mod, now used by ee10's Jetty WebSocket.
* Fixed OSGi references to old artifactIds.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-core/jetty-websocket/jetty-websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/internal/ServerUpgradeResponseImpl.java:
```diff
@@ -55,6 +56,8 @@ public String getAcceptedSubProtocol()
     {
         negotiation.setSubprotocol(protocol);
+        // Use the nested headers to avoid recursion.
+        response.getHeaders().put(HttpHeader.SEC_WEBSOCKET_SUBPROTOCOL, protocol);
```
It is not clear why this was moved from `WebScoketNegotiation.setProtocol`?
The comment is appreciated, but would be good to say who checks it and thus avoids recursing?

sbordet replied:
The comment was expanded to specify that the header was set on the _nested_ headers, not the wrapped ones.

gregw commented on jetty-core/jetty-websocket/jetty-websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketUpgradeHandler.java:
```diff
@@ -0,0 +1,60 @@
+        WebSocketUpgradeHandler wsHandler = new WebSocketUpgradeHandler(WebSocketServerComponents.ensureWebSocketComponents(server, context));
+        ServerWebSocketContainer container = wsHandler.container;
+        context.getContext().setAttribute(WebSocketContainer.class.getName(), container);
+        context.addManaged(container);
```
why is the context directly managing the container?  Surely it is managed by the WSUHandler, which itself will be managed by the context?

It is kind of strange that this method does not insert the WSUHandler into the context?

sbordet replied:
Now the container is managed by the WSUH.

Application code may want to insert another `Handler` between the `ContextHandler` and the WSUH, so it is not done here.

gregw commented on jetty-core/jetty-websocket/jetty-websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketUpgradeHandler.java:
```diff
@@ -0,0 +1,60 @@
+
+    private WebSocketUpgradeHandler(WebSocketComponents components)
+    {
+        this.container = new ServerWebSocketContainer(new WebSocketMappings(components));
```
should addBean the container here

sbordet replied:
Done.

gregw reviewed (changes requested):
A couple of initial comments.... will look more later today

sbordet committed:
Fixes #9391 - port/move Jetty WebSocket APIs, client and server to jetty-core.

* Javadocs.
* Updates after initial review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixes #9391 - port/move Jetty WebSocket APIs, client and server to jetty-core.

Added test to show how to lookup and use ServerWebSocketContainer from a Handler.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.0.x' into 'fix/jetty-12-9391-move-websocket-jetty'.

sbordet committed:
Javadocs.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.0.x' into 'fix/jetty-12-9391-move-websocket-jetty'.

sbordet committed:
Updated javadocs after review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lachlan-roberts commented on jetty-core/jetty-websocket/jetty-websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/ServerWebSocketContainer.java:
```diff
@@ -0,0 +1,278 @@
+    private final WebSocketMappings mappings;
+    private final FrameHandlerFactory factory;
+
+    ServerWebSocketContainer(WebSocketMappings mappings)
```
Any reason for this to be private. Seems the only way for this to be created is from the `WebSocketUpgradeHandler`.

There is a use case for the `JettyWebSocketServerContainer` where no `WebSocketUpgradeFilter` or `WebSocketServlet` is ever created or any mappings added, and the upgrades can still be done manually through the container with the programmatic upgrade method.

Maybe this is something we want to support here as well?

sbordet replied:
Agreed, done.

lachlan-roberts commented:
@sbordet approved, but can you respond to the review comments from me and Greg.

sbordet committed:
Made ServerWebSocketContainer usable without WebSocketUpgradeHandler,
by making the constructor and the handle() method public.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

paulrutter commented:
@sbordet i'm in the process of moving from Jetty 9 (through Apache Felix Http) to Jetty 12, by using the latest Jetty 12 Felix Http bundle (https://github.com/apache/felix-dev/tree/master/http/jetty12).

This is going well, except for the websockets part. I used Jetty Websockets in Jetty 9, which were part of the Jetty classes bundles in the Felix Http bundle, but these are no longer there in 12.

https://github.com/apache/felix-dev/blob/master/http/jetty12/pom.xml doesn't seem to include the websocket bundles. What would be the best way to approach this? Do i need a patched Felix bundle with additional dependencies on the Jetty websocket bundles? Or is there another way?

Thanks in advance.

@cziegeler for visibility.

sbordet commented:
@paulrutter you need to add the Jetty WebSocket dependencies, I think.

This should help:
https://eclipse.dev/jetty/documentation/jetty-12/programming-guide/index.html#pg-migration-11-to-12

In particular:
https://eclipse.dev/jetty/documentation/jetty-12/programming-guide/index.html#pg-websocket

Jetty jars should be OSGi bundles already. If not, report back to us (in a different issue).

You can now avoid the dependency on Jakarta EE APIs, if you want, and just use the Jetty core WebSocket APIs, see https://eclipse.dev/jetty/documentation/jetty-12/programming-guide/index.html#pg-server-websocket-jetty.

paulrutter commented:
I opened https://github.com/apache/felix-dev/pull/298 to get the websocket bundles added to felix http jetty. I got it to work this way.
